### PR TITLE
Update/stricter block validation

### PR DESCRIPTION
**Pull request by njpanderson**
This is a PR for resolution to issue #362.

### CODE REVIEW

**Comment by njpanderson** [comment]
Anyone fancying a look at this? It's regarding #362.

File: blocks/api/registration.js [hunk 1]
```diff
@@ -44,6 +44,19 @@ export function registerBlockType( name, settings ) {
 		);
 		return;
 	}
+	if ( ! settings ||
+			( settings && typeof settings.save !== 'function' ) ) {
```
Comment (youknowriad) [inline blocks/api/registration.js]: Could we write this like?

```js
if ( 
  ! settings  ||
  ( settings && typeof settings.save !== 'function' )
)
```

I think this pattern is the one used in the Gutenberg code

Comment (youknowriad) [inline blocks/api/registration.js]: Minor: should we use `isFunction` from `lodash`? It handles a special case for us.

**Comment by njpanderson** [comment]
Have amended, @youknowriad.

File: blocks/api/registration.js [hunk 2]
```diff
@@ -44,6 +49,19 @@ export function registerBlockType( name, settings ) {
 		);
 		return;
 	}
+	if ( ! settings ||
+			( settings && ! isFunction( settings.save ) ) ) {
```
Comment (youknowriad) [inline blocks/api/registration.js]: Could we put this test in one line too?

File: blocks/api/registration.js [hunk 3]
```diff
@@ -44,6 +49,19 @@ export function registerBlockType( name, settings ) {
...
+		console.error(
+			'A "save" function must be created within the block settings.'
+		);
+		return;
+	}
+	if ( settings && 'edit' in settings && ! isFunction( settings.edit ) ) {
```
Comment (youknowriad) [inline blocks/api/registration.js]: I guess the `settings` test is useless here, because we already checked for its existence above?

**Comment by njpanderson** [comment]
Both issues sorted in latest push.

**Review by youknowriad — APPROVED** [review]
LGTM 👍  Thanks for the updates

**Comment by youknowriad** [comment]
@nylen do you have any other thoughts on this. I know you're behind the validation we had in the first place.

File: blocks/api/registration.js [hunk 4]
```diff
@@ -44,6 +49,18 @@ export function registerBlockType( name, settings ) {
 		);
 		return;
 	}
+	if ( ! settings || ( settings && ! isFunction( settings.save ) ) ) {
```
Comment (nylen) [inline blocks/api/registration.js]: The `settings &&` check here is redundant - by this point in the conditional, we already know that `settings` is set.

File: blocks/api/registration.js [hunk 5]
```diff
@@ -44,6 +49,18 @@ export function registerBlockType( name, settings ) {
...
+		console.error(
+			'A "save" function must be created within the block settings.'
```
Comment (nylen) [inline blocks/api/registration.js]: Let's clean up this message a bit:

> The "save" property must be specified and must be a valid function.

File: blocks/api/test/factory.js [hunk 6]
```diff
@@ -10,6 +11,8 @@ import { createBlock, switchToBlockType } from '../factory';
 import { getBlockTypes, unregisterBlockType, setUnknownTypeHandler, registerBlockType } from '../registration';
 
 describe( 'block factory', () => {
+	const defaultBlockType = { save: noop };
```
Comment (nylen) [inline blocks/api/test/factory.js]: `defaultBlockSettings` would be a better name here and below.

**Review by nylen — CHANGES_REQUESTED** [review]
Generally seems fine to me. Some notes above. We should also make sure similar validation is in place on the server side, but for now this would be for the `render` function.

**Comment by nylen** [comment]
> We should also make sure similar validation is in place on the server side, but for now this would be for the `render` function.

To clarify: this is fine for a separate PR.

**Comment by njpanderson** [comment]
@nylen / @youknowriad - Both changes above have been made.

**Review by nylen — APPROVED** [review]
Looks good to me. Thanks for working on this!

**Comment by swissspidy** [comment]
Note that because of this there's a bunch of warnings when running the unit tests.

**Comment by swissspidy** [comment]
Can de `validate-block-settings` branch be deleted?

**Comment by njpanderson** [comment]
Apologies - I've been quite tied up for a few days.

Branch has been deleted.

In answer to your remark regarding warnings, @swissspidy - I'd fixed all of them on the tip of my branch but I suspect due to the lead time more old patterns of use for the registration have crept in. Not sure who's down for fixing them but happy to work on another issue for it if there is one.

**Comment by nylen** [comment]
These warnings should really be errors that fail the build, which will prevent them from reoccurring.  See #1769.

**Pull request merged**